File: eio2024-lv/2024-10-07-lv/voistlus/solution/sol.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int main()
{
    long long mod = 1e9+7;
    int n;
    cin >> n;
    int i, j, k, l;
    int a[n];
    int b[n];
    for (i = 0; i < n; i++) {
        cin >> a[i];
    }
    for (i = 0; i < n; i++) {
        cin >> b[i];
    }
    sort(b, b + n, greater<int>());
    for (i = 0; i < n; i++) {
        long long ans = 0;
        for (j = 0; j < n; j++) {
            long long p = 1;
            int s = a[i] + b[j];
            l = 0;
            for (k = 0; k < n; k++) {
                if (k == i) {
                    continue;
                }
                while (l < n && b[n - l - 1] + a[k] <= s) {
                    l++;
                }
                p *= l - (l >= n - j) - k + (k > i);
                if (p <= 0) {
                    p = 0;
                    break;
                }
                p %= mod;
            }
            ans += p;
            ans %= mod;
        }
        cout << ans << endl;
    }
}

File: eio2024-lv/2024-10-07-lv/voistlus/solution/sol.py
mod = 1000000007
n = int(input())
a = list(map(int, input().split()))
b = list(map(int, input().split()))
b.sort(reverse=True)
for i in range(n):
    ans = 0
    for j in range(n):
        p = 1
        s = a[i] + b[j]
        l = 0
        for k in range(n):
            if k == i:
                continue
            while l < n and b[n - l - 1] + a[k] <= s:
                l += 1
            p *= l - (l >= n - j) - k + (k > i);
            if p <= 0:
                p = 0
                break
            p %= mod
        ans += p
        ans %= mod
    print(ans)
